src/Symfony/Component/Security/Guard/AuthenticatorInterface.php

```diff
@@ -90,14 +90,15 @@ public function getUser($credentials, UserProviderInterface $userProvider);
      *
      * The *credentials* are the return value from getCredentials()
      *
-     * @param mixed         $credentials
-     * @param UserInterface $user
+     * @param mixed                     $credentials
+     * @param UserInterface             $user
+     * @param PasswordUpgraderInterface $passwordUpgrader
      *
      * @return bool
      *
      * @throws AuthenticationException
      */
-    public function checkCredentials($credentials, UserInterface $user);
+    public function checkCredentials($credentials, UserInterface $user/*, PasswordUpgraderInterface $passwordUpgrader = null*/);
```

Comment:
This interface is not aware of the password concept so far, which is good since not all authenticators deal with passwords, see e.g. the one used in the docs https://symfony.com/doc/current/security/guard_authentication.html#step-2-create-the-authenticator-class.
It's also one of the goals of https://github.com/symfony/symfony/issues/30914 to remove the need for writing a no-op `getPassword()` for cases where authentication is not password-based, as such cases are becoming more and more common.

Verifying a password in a guard authenticator requires to give it a user password encoder via constructor/setter injection and to write the password verification logic. I don't think it should be different for password migration.

Reply:
I'm sorry but I don't understand what you suggest.
If you have a better idea, could you please provide some guidance?

Reply:
i think the migration should be part of `getUser()` here, which already receives the user provider/password upgrader.

Strictly spoken "checking credentials" does not really mean "upgrading a password" right

Reply:
Meaning the user is responsible to implement the migration for authenticators as needed. I agree with @chalasr, we dont know from the contract if we're dealing with a type of password authentication

Reply:
That's why the argument is nullable.
And also why it's called *opportunistic* password migration.
To my understanding of the code, this PR implements this in the most sensible way.
"Responsibility" is not something that comes out from thin air but emerges from requirements.
And requirements are clear: when checking a password, upgrade it while doing so if it provides an improvement.

Reply:
But please, do send a PR if you have another proposal!

Reply:
hm actually you're right. You can only migrate the password after checking it

Reply:
Dealing with a password or not is implementation specific, this dependency should be an implementation detail to me, same as the password encoder needed to verify it, and so should not be part of this interface. If we had e.g. built-in JWT support, should we also update this interface to give it a token refresher? It's too much opinionated for this particular interface IMO, nullable or not.

Ideally, we would just implement `checkCredentials()` in `AbstractFormLoginAuthenticator` and make it handle both password verification and migration, letting this interface as it is. Could eventually be specified by a sub-interface meant to add password verification+migration capabilities to this one, targeting password-based implementations.
That's a blocker to me, I'd prefer to not add such capabilities to guard if doing so means coupling it with the password concept, as it is totally free of it currently.

Reply:
@chalasr i gave it another attempt after discussing with @nicolas-grekas , see https://github.com/nicolas-grekas/symfony/pull/23

Reply:
Thank you @ro0NL it's been definitely useful.
PR now embeds a new `Guard\PasswordAuthenticatedInterface` with a single `getPassword()` method, see https://github.com/symfony/symfony/compare/e319f545c3488aaa2d0008efb41889b84175730f..067d4fb8fca14ef15c18171fa075e77f3c1c39e7